tsdb/exemplar.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/prometheus/prometheus/pkg/exemplar"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+type exemplarMetrics struct {
+	outOfOrderExemplars prometheus.Counter
+	duplicateExemplars  prometheus.Counter
+}
+
+func newExemplarMetrics(r prometheus.Registerer) *exemplarMetrics {
+	m := &exemplarMetrics{
+		outOfOrderExemplars: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prometheus_exemplar_out_of_order_exemplars_total",
+			Help: "Total number of out of order samples ingestion failed attempts",
+		}),
+		duplicateExemplars: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prometheus_exemplar_duplicate_exemplars_total",
+			Help: "Total number of series in the head block.",
+		}),
+	}
+	if r != nil {
+		r.MustRegister(
+			m.outOfOrderExemplars,
+			m.duplicateExemplars,
+		)
+	}
+	return m
+}
+
+type CircularExemplarStorage struct {
+	metrics   *exemplarMetrics
+	lock      sync.RWMutex
+	index     map[string]indexEntry
+	exemplars []*circularBufferEntry
+	nextIndex int
+}
+
+type indexEntry struct {
+	first int
+	last  int
+}
+
+type circularBufferEntry struct {
+	exemplar     exemplar.Exemplar
+	seriesLabels labels.Labels
+	next         int
+}
+
+// If we assume the average case 95 bytes per exemplar we can fit 5651272 exemplars in
+// 1GB of extra memory, accounting for the fact that this is heap allocated space.
+func NewCircularExemplarStorage(len int, reg prometheus.Registerer) (*CircularExemplarStorage, error) {
+	if len < 1 {
+		return nil, errors.Errorf("must initialize exemplar storage with space for at least 1 exemplar, %d is invalid", len)
+	}
+	return &CircularExemplarStorage{
+		exemplars: make([]*circularBufferEntry, len),
+		index:     make(map[string]indexEntry),
+		metrics:   newExemplarMetrics(reg),
+	}, nil
+}
+
+func (ce *CircularExemplarStorage) Appender() storage.ExemplarAppender {
+	return ce
+}
+
+func (ce *CircularExemplarStorage) ExemplarAppender() storage.ExemplarAppender {
+	return ce
+}
+
+func (ce *CircularExemplarStorage) ExemplarQuerier(_ context.Context) (storage.ExemplarQuerier, error) {
+	return ce, nil
+}
+
+func (ce *CircularExemplarStorage) Querier(ctx context.Context) (storage.ExemplarQuerier, error) {
+	return ce, nil
+}
+
+// Select returns exemplars for a given set of series labels hash.
+func (ce *CircularExemplarStorage) Select(start, end int64, l labels.Labels) ([]exemplar.Exemplar, error) {
+	var (
+		ret []exemplar.Exemplar
+		e   *circularBufferEntry
+		idx indexEntry
+		ok  bool
+	)
+
+	ce.lock.RLock()
+	defer ce.lock.RUnlock()
+
+	if idx, ok = ce.index[l.String()]; !ok {
+		return nil, nil
+	}
+
+	e = ce.exemplars[idx.first]
+	for {
+		if e.exemplar.Ts < start {
+			if e.next == -1 {
+				break
+			}
+			e = ce.exemplars[e.next]
+			continue
+		}
+		if e.exemplar.Ts > end {
+			break
+		}
+
+		ret = append(ret, e.exemplar)
+		if e.next == -1 {
+			break
+		}
+		e = ce.exemplars[e.next]
+	}
```

Comment:
How about this?
```go
for e = ce.exemplars[idx.first]; e.next != -1 && e.exemplar.Ts <= end; e = ce.exemplars[e.next] {
	if e.exemplar.Ts >= start || e.exemplar.Ts <= end {
		ret = append(ret, e.exemplar)
	}
}
```

Reply:
This wouldn't be entirely correct, since we only want to skip an exemplar when `e.next == -1` if the timestamp is before `start`, if it's timestamp is greater than start and less than end we should still include it in the result. Overall I'd really prefer to go with more LoC for readability here.